### DISCUSSION TITLE
tastmota32: USE_SCRIPT triggers xdrv_10_scripter.ino:72:22: fatal error: renderer.h: No such file or directory

### DISCUSSION THREAD

**Discussion by corbolais** [arendst/Tasmota#12864] [Q&A] [answered]
I want to use scripting in tasmota32 on a `ai thinker esp32-cam` module.

`#define USE_SCRIPT` makes `pio run -e tasmota32-webcam` fail in branch master v9.5.0:
```
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32-cam.html
PLATFORM: Espressif 32 (3.2.0) > AI Thinker ESP32-CAM, 4M Flash 4MB PSRAM, Tasmota 1856k Code/OTA, 320k SPIFFS
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (esp-prog) External (esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 3.10006.210326 (1.0.6) 
 - tool-esptoolpy 1.30000.201119 (3.0.0) 
 - tool-mklittlefs 1.203.210628 (2.3) 
 - toolchain-xtensa32 2.50200.97 (5.2.0)
 ```
[...]
```
Compiling .pio/build/tasmota32-webcam/src/tasmota.ino.cpp.o
tasmota/Tasmota/tasmota/xdrv_10_scripter.ino:72:22: fatal error: renderer.h: No such file or directory

```

This used to work(tm) before.

BTW, Espressif 32 (3.2.0) is used to compile although it just got updated to 3.3.1 with a `pio update`. How to enforce the new version?

Ideas anyone?

**Comment by corbolais** [comment]
> BTW, Espressif 32 (3.2.0) is used to compile although it just got updated to 3.3.1 with a `pio update`. How to enforce the new version?

Section `[core32]` in `platformio_tasmota32.ini`, edit:

`
platform                    = espressif32 @ 3.3.1
`

**Comment by corbolais** [comment]
Is there a way to rotate the camera picture in tasmota?

**Reply by ascillato** [reply]
No. There isn't. Sorry.

**Reply by corbolais** [reply]
What about a CSS rotation like this:

`<img id="image_canv" src="/image.png" class="rotate90">`

and 

```
.rotate90 {
    -webkit-transform: rotate(90deg);
    -moz-transform: rotate(90deg);
    -o-transform: rotate(90deg);
    -ms-transform: rotate(90deg);
    transform: rotate(90deg);
}
```

As in [here](https://stackoverflow.com/questions/20061774/rotate-an-image-in-image-source-in-html).
Maybe configurable via webui.

**Reply by sfromis** [reply]
Such a HTML rotate would IMO be of quite limited utility, as the intended main usage of webcam is that it can provide streaming to some sort of monitoring system, not to use the barebones test/config web UI.

**Reply by corbolais** [reply]
Yes, that's correct. It'd be only useful for still images.

**Comment by barbudor** [comment, marked as answer]
You can't add SCRIPT to tasmota32-webcam as it is fine-tuned build.
As per Tasmota docs, you should never add any feature to a pre-defined variant which are all fine-tuned.
You must derive your own variant from tasmota/tasmota32 and add what is necessary. This means adding both Webcam and script to tasmota32.
This has already been addressed multiple times in these Discussions so please search

**Reply by corbolais** [reply]
OK, thank you. Could you elaborate on how to add the library path so the build process might find it?

**Reply by barbudor** [reply]
That's the point, tasmota32 already has all the libs so you don't need to add libraries.

**Reply by corbolais** [reply]
Hmm. Still don't get it. Why's it failing for a missing header file then?

**Reply by barbudor** [reply]
Clean way
- Copy `platformio_tasmota_cenv_sample.ini` into `platformio_tasmota_cenv.ini`
- Add a new environment section same as tasmota32-webcam but without the `lib_extra_dirs` line
```
[env:tasmota32-webcam-allibs]
extends                 = env:tasmota32_base
board                   = esp32-cam
build_flags             = ${env:tasmota32_base.build_flags} -DFIRMWARE_WEBCAM
```
- Add in your `user_config_override.h`
```
#undef USE_RULES
#define USE_SCRIPT
```
- Build `tasmota32-webcal-alllibs`

This way, your custom environment ("cenv") will not be overwritten when pulling Tasmota updates.

Lazy way:
- Just comment out the line `lib_extra_dirs` of environment `tasmota32-webcam` 
But this will be overwritten every time you pull Tasmota updates.

**Comment by Jason2866** [comment]
I suggest to use the development branch to compile your own build. In the development branch there are enhancements in the used core (we use our own Tasmota optimized version) and some bug fixes in Tasmota.
